### ISSUE TITLE
Can quote-props: as-needed respect camelCase with properties: always?

### ISSUE THREAD

**Issue by csciubbach** [eslint/eslint#13053]
**What rule do you want to change?**
quote-props
dot-notation

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
Preferably new default. If we have
```json
"camelcase": ["error", {
    "properties": "always"
}],
```
Then quote-props as-needed should make it okay and in fact force you to quote non camel case properties.

**Please provide some example code that this change will affect:**

```js
const whatever = {
    id: 123,
    snake_case_word: "test"
};
```

**What does the rule currently do for this code?**
Allows it, and if you put quotes rules on it complains the quotes are not required, even though it is not a valid property name if you have camelcase checking properties. Essentially you have to turn off one of these rules which is a bad compromise.

**What will the rule do after it's changed?**
Allow you to have both camelCase properties rule and as-needed quote props work to enforce quotes on snake_case non camelCase propertis IF camelCase is on and properties is on.

**Are you willing to submit a pull request to implement this change?**
Maybe

**Comment by csciubbach** [comment]
dot-notation should respect this also, and not suggest dot-notation for a property which violates camelCase

**Comment by anikethsaha** [comment]
just to summarise about your issue 
when the code is

[`demo`](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgcXVvdGUtcHJvcHM6IFtcImVycm9yXCIsIFwiYXMtbmVlZGVkXCJdKi9cbi8qZXNsaW50IGNhbWVsY2FzZTogW1wiZXJyb3JcIl0qL1xuXG5jb25zdCB3aGF0ZXZlciA9IHtcbiAgICBpZDogMTIzLFxuICAgIHNuYWtlX2Nhc2Vfd29yZDogXCJ0ZXN0XCJcbn07Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7InF1b3RlLXByb3BzIjoyLCJjYW1lbGNhc2UiOjJ9LCJlbnYiOnt9fX0=)
```js
/*eslint quote-props: ["error", "as-needed"]*/
/*eslint camelcase: ["error"]*/

const whatever = {
    id: 123,
    snake_case_word: "test" // camelcase rule error
};
```

and with quotes [`demo`](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgcXVvdGUtcHJvcHM6IFtcImVycm9yXCIsIFwiYXMtbmVlZGVkXCJdKi9cbi8qZXNsaW50IGNhbWVsY2FzZTogW1wiZXJyb3JcIl0qL1xuXG5jb25zdCB3aGF0ZXZlciA9IHtcbiAgICBpZDogMTIzLFxuICAgICdzbmFrZV9jYXNlX3dvcmQnOiBcInRlc3RcIlxufTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsicXVvdGUtcHJvcHMiOjIsImNhbWVsY2FzZSI6Mn0sImVudiI6e319fQ==)

```js
/*eslint quote-props: ["error", "as-needed"]*/
/*eslint camelcase: ["error"]*/

const whatever = {
    id: 123,
    "snake_case_word": "test" // quote-prop rule error
};
```

so there is an inconsistent between these two rules
```js
/*eslint quote-props: ["error", "as-needed"]*/
/*eslint camelcase: ["error"]*/
```

right ?

 in this case, you can try using `/*eslint camelcase: ["error",{"properties": "never"}]*/`, it doesnt enforce camelcase for property names.  

[`demo`](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgcXVvdGUtcHJvcHM6IFtcImVycm9yXCIsIFwiYXMtbmVlZGVkXCJdKi9cbi8qZXNsaW50IGNhbWVsY2FzZTogW1wiZXJyb3JcIix7XCJwcm9wZXJ0aWVzXCI6IFwibmV2ZXJcIn1dKi9cblxuY29uc3Qgd2hhdGV2ZXIgPSB7XG4gICAgaWQ6IDEyMyxcbiAgICBzbmFrZV9jYXNlX3dvcmQ6IFwidGVzdFwiXG59O1xuXG5jb25zdCBhc2RfYXNkYXNkID0gMTAiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsicXVvdGUtcHJvcHMiOjIsImNhbWVsY2FzZSI6Mn0sImVudiI6e319fQ==)

**Comment by mdjermanovic** [comment]
@csciubbach thanks for the issue!

Rules are, by design, unaware of other rules' configuration, so this would have to be an additional option in the `quote-props` rule.

As @anikethsaha mentioned, [allowPattern](https://eslint.org/docs/rules/dot-notation#allowpattern) in `dot-notation` is particularly designed for this use case.

A similar option in `quote-props` seems reasonable to me :+1:

**Comment by anikethsaha** [comment]
> A similar option in `quote-props` seems reasonable to me 👍

Yes , this is much wider solution than a specific option for just camelcase. 


+1 for `allowPattern` in `quote-props`. Lets hear from others 👍
